hook_test.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"bytes"
+	"regexp"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHookAddCaller(t *testing.T) {
+	buf := &bytes.Buffer{}
+	logger := NewJSON(All, Output(buf), AddCaller())
+	logger.Info("Callers.")
+
+	re := regexp.MustCompile(`hook_test.go:[\d]+: Callers\.`)
+	assert.Regexp(t, re, buf.String(), "Expected to find package name and file name in output.")
+}
+
+func TestHookAddCallerFail(t *testing.T) {
+	buf := &bytes.Buffer{}
+	errBuf := &bytes.Buffer{}
+
+	originalSkip := _callerSkip
+	_callerSkip = 1e3
+	defer func() { _callerSkip = originalSkip }()
+
+	logger := NewJSON(All, Output(buf), ErrorOutput(errBuf), AddCaller())
+	logger.Info("Failure.")
+	assert.Equal(t, "failed to get caller\n", errBuf.String(), "Didn't find expected failure message.")
+	assert.Contains(t, buf.String(), `"msg":"Failure."`, "Expected original message to survive failures in runtime.Caller.")
+}
+
+func TestHookAddStacks(t *testing.T) {
```

Comment:
i think we want to test both branches in `AddStacks`, where the level is higher than or equal to the minimum, and one for when it's lower. Maybe a simple table test that takes a logger configured for `Warning` and ensures `Warning` & `Error` add the stack, while `Info` doesn't?